staging/src/k8s.io/kubectl/pkg/kuberc/kuberc.go

```diff
@@ -0,0 +1,303 @@
+// alias that is currently executed from args. After that it sets the flag definitions in alias as default values
+// of the command. Lastly, others parameters (e.g. resources, etc.) that are passed as arguments in kuberc
+// is appended into the os.Args.
+func (p *Preferences) applyAliases(rootCmd *cobra.Command, kuberc *v1alpha1.Preferences, args []string, errOut io.Writer) ([]string, error) {
+	_, _, err := rootCmd.Find(args[1:])
+	if err == nil {
+		// Command is found, no need to continue for aliasing
+		return args, nil
+	}
+
+	aliasArgsMap := make(map[string]struct {
+		args    []string
+		flags   []v1alpha1.CommandOverrideFlag
+		command *cobra.Command
+	})
+
+	for _, alias := range kuberc.Aliases {
+		// do not allow shadowing built-ins
+		if _, _, err := rootCmd.Find([]string{alias.Name}); err == nil {
+			fmt.Fprintf(errOut, "Warning: Setting alias %q to a built-in command is not supported\n", alias.Name)
+			continue
+		}
+
+		if _, ok := aliasArgsMap[alias.Name]; ok {
+			fmt.Fprintf(errOut, "alias %s is already set, skipping...\n", alias.Name)
+			continue
+		}
+
+		commands := strings.Split(alias.Command, " ")
+		existingCmd, flags, err := rootCmd.Find(commands)
+		if err != nil {
+			fmt.Fprintf(errOut, "Warning: command %q not found to set alias %q: %v\n", alias.Command, alias.Name, flags)
+			continue
+		}
+
+		newCmd := *existingCmd
+		newCmd.Use = alias.Name
+		aliasCmd := &newCmd
+
+		aliasArgsMap[alias.Name] = struct {
+			args    []string
+			flags   []v1alpha1.CommandOverrideFlag
+			command *cobra.Command
+		}{
+			args:    alias.Args,
+			flags:   alias.Flags,
+			command: aliasCmd,
+		}
+	}
+
+	// It is verified that all the aliases are valid. So that, now we
+	// can define them in to the root command.
+	for key, val := range aliasArgsMap {
+		p.aliases[key] = struct{}{}
+		rootCmd.AddCommand(val.command)
+	}
+
+	var aliasName string // first "non-flag" arguments
+	for _, arg := range args[1:] {
+		if !strings.HasPrefix(arg, "-") {
+			aliasName = arg
+			break
+		}
+	}
+
+	foundAliasCmd, _, err := rootCmd.Find([]string{aliasName})
+	if err != nil {
+		return args, nil
+	}
+
+	aliasArgs, ok := aliasArgsMap[aliasName]
+	if !ok {
+		return args, nil
+	}
+
+	// This function triggers merging the persistent flags in the parent commands.
+	_ = foundAliasCmd.InheritedFlags()
+
+	for _, fl := range aliasArgs.flags {
+		err = foundAliasCmd.Flags().Set(fl.Name, fl.Default)
+		if err != nil {
+			return args, fmt.Errorf("could not apply value %s to flag %s in alias %s err: %w", fl.Default, fl.Name, args[0], err)
+		}
+	}
+
+	// all args defined in kuberc should be appended to actual args.
+	args = append(args, aliasArgs.args...)
+	// Cobra (command.go#L1078) appends only root command's args into the actual args and ignores the others.
+	// We are appending the additional args defined in kuberc in here and
+	// expect that it will be passed along to the actual command.
+	rootCmd.SetArgs(args[1:])
```

Comment:
I think this piece of code is the problem behind what I described in https://github.com/kubernetes/kubernetes/pull/125230#issuecomment-2208709072.

Reply:
Yes, you are right. Probably we should reverse the order in here https://github.com/kubernetes/kubernetes/pull/125230/files#diff-5d7e607393e8973d94e9acfa8c87890b971c94a5381d758d400d73e3dbe2d598R232 because it is unlikely that such usage:

```yaml
apiVersion: kubectl.config.k8s.io/v1alpha1
kind: Preferences
aliases:
- name: getns
  command: get
  args:
   - default
```
and 

```sh
$ KUBECTL_KUBERC=true kubectl getns namespace 
```

Reply:
Yeah, I'd basically expect alias to handle the initial bits, and everything else is always appended at the end.

Reply:
I spent some time on this topic today. Supporting this functionality is not only complicating the code but more importantly would be impractical from user's point of view.

### Current Behavior

Although there is a partial control in the args of kuberc just like:

```yaml
apiVersion: kubectl.config.k8s.io/v1alpha1
kind: Preferences
aliases:
- name: getns
  command: get
  args:
   - namespace
```
command arguments are very arbitrary, for instance all these command executions are valid:

```sh
$ KUBECTL_KUBERC=true kubectl getns --some-flag=value --some-flag2 value2 namespace default 
$ KUBECTL_KUBERC=true kubectl getns  namespace default  --some-flag=value --some-flag2 value2
$ KUBECTL_KUBERC=true kubectl getns --some-flag=value  namespace/default   --some-flag2 value2
```
Basically with this ordering `args = append(args, aliasArgs.args...)`, we don't support none of them ^ in this kuberc definition. Instead we recommend to define full arguments path as a first option:

```yaml
apiVersion: kubectl.config.k8s.io/v1alpha1
kind: Preferences
aliases:
- name: getns
  command: get
  args:
   - namespace
   - default
```
or if user doesn't want to stick to default namespace, remove args entirely and use it dynamically:

```yaml
apiVersion: kubectl.config.k8s.io/v1alpha1
kind: Preferences
aliases:
- name: getns
  command: get
```

```sh
$ KUBECTL_KUBERC=true kubectl getns namespace whatever
```

### Reversing the order by appending actual args into kuberc args

Let's take the same example kuberc:

```yaml
apiVersion: kubectl.config.k8s.io/v1alpha1
kind: Preferences
aliases:
- name: getns
  command: get
  args:
   - namespace
```
and this command execution

```sh
$ KUBECTL_KUBERC=true kubectl getns --some-flag=value --some-flag2 value2 default 
```

in this format actual args list is in this:

```go
[]string [
	"kubectl",
	"getns",
         "--some-flag=value",
         "--some-flag2",
         "value2",
         "default"
]
```
appending actual arguments tail of kuberc arguments causes this:

```go
[]string [
        "namespace",
	"kubectl",
	"getns",
         "--some-flag=value",
         "--some-flag2",
         "value2",
         "default"
]
``` 
that is clearly corrupted command execution because there is no command name is `namespace`.

Let's say we find the command paths and all the flag definitions (that would be error prone due to arbitrary usage) and add `namespace` argument after them but before the actual arguments (just before the `default`). I mean something like this.

```sh
$ KUBECTL_KUBERC=true kubectl getns --some-flag=value --some-flag2 value2 (adding namespace in here) default
```
That would work. But what about this usage (default is in the middle of arbitrary flag definitions):

```sh
$ KUBECTL_KUBERC=true kubectl getns --some-flag value  (finding here and add namespace) default --some-flag2 value2
```

We always take single arguments as an example but when multiple arguments come in, that become very complicated and from user's point of view, it is hard to memorize the current behavior. 

For example, what user should expect this definition:

```yaml
apiVersion: kubectl.config.k8s.io/v1alpha1
kind: Preferences
aliases:
- name: getns
  command: get
  args:
   - namespace
   - somenamespace
```

```sh
$ KUBECTL_KUBERC=true kubectl getns --some-flag=value --some-flag2 value2 default
```

If we force that kuberc arguments will be appended at the end of actual args, user can easily infer that this would lead to this incorrect command:

```sh
$ KUBECTL_KUBERC=true kubectl getns --some-flag=value --some-flag2 value2 default namespace somenamespace
```
otherwise, in my opinion, it is hard to say the expected behavior.

Reply:
That's definitely something we might want to think more later on. Like I said earlier, it's not a blocker for this PR.